src/util/positioning.ts

```diff
@@ -214,42 +77,68 @@ export function positionElements(
     });
   }
 
-  // coordinates where to position
-
-  // Required for transform:
-  const style = targetElement.style;
-  style.position = 'absolute';
-  style.top = '0';
-  style.left = '0';
-  style['will-change'] = 'transform';
-
-  let testPlacement: Placement | null = null;
-  let isInViewport = false;
-  for (testPlacement of placementVals) {
-    let addedClasses = addClassesToTarget(testPlacement);
-
-    if (positionService.positionElements(hostElement, targetElement, testPlacement, appendToBody)) {
-      isInViewport = true;
-      break;
-    }
-
-    // Remove the baseClasses for further calculation
-    if (baseClass) {
-      addedClasses.forEach((classname) => { classList.remove(classname); });
-    }
-  }
+  const popperPlacements = placementVals.map((_placement) => { return convertPlacementToPopper(_placement); });
 
-  if (!isInViewport) {
-    // If nothing match, the first placement is the default one
-    testPlacement = placementVals[0];
-    addClassesToTarget(testPlacement);
-    positionService.positionElements(hostElement, targetElement, testPlacement, appendToBody);
-  }
+  let mainPlacement = popperPlacements.shift();
+
+  const bsModifier: Partial<Modifier<any, any>> = {
+    name: 'bootstrapClasses',
+    enabled: true,
+    phase: 'write',
+    fn({state}) {
+      if (baseClass) {
+        const bsClassRegExp = new RegExp(baseClass + '-[a-z]+', 'gi');
+
+        const popperElement: HTMLElement = state.elements.popper as HTMLElement;
+        const popperPlacement = state.placement;
+
+        let className = popperElement.className;
+
+        // Remove old bootstrap classes
+        className = className.replace(bsClassRegExp, '');
+
+        // Add current placements
+        className += ` ${convertPlacementToBsClasses(baseClass, popperPlacement)}`;
+
+        // Remove multiple spaces
+        className = className.trim().replace(spacesRegExp, ' ');
+
+        // Reassign
+        popperElement.className = className;
+      }
+    },
+  };
 
-  return testPlacement;
+  const popperInstance = createPopper(hostElement, targetElement, {
+    placement: mainPlacement,
+    modifiers: [
+      bsModifier,
+      {
+        name: 'offset',
+        enabled: true,
+        options: {
+          offset: () => {
+            return baseClass === 'bs-popover' ? [0, 12] : [];
```

Comment:
This should be passed as a partial config only from calling `positionElements` from the popover...
We should not have component-specific code here. Also 8px, not 12 → https://github.com/twbs/bootstrap/blob/main/js/src/popover.js#L25